### PR TITLE
Fixed CertUtils.handleOtherKeys() behavior with Cavium

### LINKED ISSUE #4509: Resource leak when creating private keys
### Describe the bug

We're frequently creating and closing `KubernetesClient` instances in our code, but it appears there's some kind of resource leak associated with private keys, leading to:

```
Caused by: com.cavium.cfm2.CFM2Exception: A call to the API importRSAPrivateKey from KeySpec failed with error code ab : HSM Error: Reached the limit of Max number of supported objects
```

Stack trace:

```
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
    at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:103)
    at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:97)
    at io.fabric8.kubernetes.client.utils.HttpClientUtils.applyCommonConfiguration(HttpClientUtils.java:223)
    at io.fabric8.kubernetes.client.okhttp.OkHttpClientFactory.createHttpClient(OkHttpClientFactory.java:96)
    at io.fabric8.kubernetes.client.okhttp.OkHttpClientFactory.createHttpClient(OkHttpClientFactory.java:32)
    at io.fabric8.kubernetes.client.utils.HttpClientUtils.createHttpClient(HttpClientUtils.java:174)
    at io.fabric8.kubernetes.client.DefaultKubernetesClient.<init>(DefaultKubernetesClient.java:153)
    ... 48 more
Caused by: java.security.spec.InvalidKeySpecException: com.cavium.cfm2.CFM2Exception: A call to the API importRSAPrivateKey from KeySpec failed with error code ab : HSM Error: Reached the limit of Max number of supported objects
    at com.cavium.key.factory.CaviumRSAKeyFactory.engineGeneratePrivate(CaviumRSAKeyFactory.java:119)
    at java.base/java.security.KeyFactory.generatePrivate(KeyFactory.java:384)
    at io.fabric8.kubernetes.client.internal.CertUtils.handleOtherKeys(CertUtils.java:177)
    at io.fabric8.kubernetes.client.internal.CertUtils.loadKey(CertUtils.java:136)
    at io.fabric8.kubernetes.client.internal.CertUtils.createKeyStore(CertUtils.java:112)
    at io.fabric8.kubernetes.client.internal.CertUtils.createKeyStore(CertUtils.java:247)
    at io.fabric8.kubernetes.client.internal.SSLUtils.keyManagers(SSLUtils.java:168)
    at io.fabric8.kubernetes.client.internal.SSLUtils.keyManagers(SSLUtils.java:157)
    at io.fabric8.kubernetes.client.utils.HttpClientUtils.applyCommonConfiguration(HttpClientUtils.java:214)
    ... 52 more
Caused by: com.cavium.cfm2.CFM2Exception: A call to the API importRSAPrivateKey from KeySpec failed with error code ab : HSM Error: Reached the limit of Max number of supported objects
    at com.cavium.cfm2.Util.handleCfm2Failure(Util.java:410)
    at com.cavium.cfm2.ImportKey.importRSAPrivateKey(ImportKey.java:315)
    at com.cavium.key.factory.CaviumRSAKeyFactory.engineGeneratePrivate(CaviumRSAKeyFactory.java:117)
    ... 60 more
```

### Fabric8 Kubernetes Client version

6.0.0

### Steps to reproduce

1. Use CaviumRSAKeyFactory
2. Run on EC2 instance with CloudHSM
3. In a loop: create then close KubernetesClients connecting to a single cluster.

### Expected behavior

Clients can be created indefinitely.

### Runtime

Kubernetes (vanilla)

### Kubernetes API Server version

1.23

### Environment

Linux

### Fabric8 Kubernetes Client Logs

```shell
Not sure what additional logs I can provide here.
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sfc-gh-jkowalski**
## Description
Fixes #4509

This took a while to find the root cause:

The internal SPI fallback logic inside `KeyFactory.generatePrivate()` has the weird side effect of latching onto the LAST registered provider (which in our case was Cavium) after `InvalidKeySpecException` is thrown.

This choice is sticky for a single instance of KeyFactory and the fix for our issue is to get fresh `KeyFactory` instance when retrying.

## Type of change

 - [x] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/master/CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [x] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by sfc-gh-dbrown** [comment]
Here is a simple repro showing that re-using the KeyFactory after an exception results in a different Provider than using a fresh KeyFactory:

```
import java.security.KeyFactory;
import java.security.Provider;
import java.security.Security;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.PKCS8EncodedKeySpec;
import java.util.Base64;

public class Repro {
    public static void main(String[] args) throws Exception {
        System.out.println("List Providers that implement KeyFactory.RSA:");
        for (Provider p : Security.getProviders()) {
            if (p.getService("KeyFactory", "RSA") != null) {
                System.out.println(" " + p.getName());
            }
        }
        KeyFactory kf = KeyFactory.getInstance("RSA");
        // Uncommenting this will change behavior as calls to getProvider() pin the provider
        //System.out.println("Provider: " + kf.getProvider().getName());

        try {
            kf.generatePrivate(new PKCS8EncodedKeySpec(invalidPKCS8));
            System.out.println("success=bad");
        } catch (InvalidKeySpecException ex) {
            System.out.println("invalid key=good");
            var key = kf.generatePrivate(new PKCS8EncodedKeySpec(validPKCS8));
            System.out.println("Provider: " + kf.getProvider().getName());

            System.out.println("Do it again with a fresh KeyFactory and get a different Provider.");
            kf = KeyFactory.getInstance("RSA");
            key = kf.generatePrivate(new PKCS8EncodedKeySpec(validPKCS8));
            System.out.println("Provider: " + kf.getProvider().getName());
        }
    }

    private static final byte[] invalidPKCS8 = new byte[] {0,0,0};

    private static final byte[] validPKCS8 =
            Base64.getDecoder()
                    .decode(
                            "MIIBVgIBADANBgkqhkiG9w0BAQEFAASCAUAwggE8AgEAAkEAq7BFUpkGp3+LQmlQYx2eqzDV+xeG8kx/sQFV18S5JhzGeIJNA72wSeukEPojtqUyX2J0CciPBh7eqclQ2zpAswIDAQABAkAgisq4+zRdrzkwH1ITV1vpytnkO/NiHcnePQiOW0VUybPyHoGM/jf75C5xET7ZQpBe5kx5VHsPZj0CBb3b+wSRAiEA2mPWCBytosIU/ODRfq6EiV04lt6waE7I2uSPqIC20LcCIQDJQYIHQII+3YaPqyhGgqMexuuuGx+lDKD6/Fu/JwPb5QIhAKthiYcYKlL9h8bjDsQhZDUACPasjzdsDEdq8inDyLOFAiEAmCr/tZwA3qeAZoBzI10DGPIuoKXBd3nk/eBxPkaxlEECIQCNymjsoI7GldtujVnr1qT+3yedLfHKsrDVjIT3LsvTqw==");
}
```

**Comment by sfc-gh-jkowalski** [comment]
The style violation is pre-existing, not sure if I should be fixing it as part of the PR.

**Comment by shawkins** [comment]
@sfc-gh-jkowalski just run mvn spotless:apply then commit those changes.  There was an outstanding issue / pr to bulk format the remaining files in the codebase, but it hasn't been committed.

**Comment by sfc-gh-jkowalski** [comment]
@shawkins done.

**Comment by sfc-gh-jkowalski** [comment]
So this style change now triggered code coverage warning. Could this PR be merged as-is or do you want me to do something about code coverage?

**Review by shawkins — APPROVED** [review]
LGTM

Don't worry about the coverage report, your core change is straight-forward.

**Comment by sfc-gh-jkowalski** [comment]
Is there anything blocking merging this PR?

File: kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/internal/CertUtils.java [hunk 1]
```diff
@@ -150,37 +156,34 @@ public PrivateKey call() {
...
-    KeyFactory keyFactory = KeyFactory.getInstance(clientKeyAlgo);
     try {
       // First let's try PKCS8
-      return keyFactory.generatePrivate(new PKCS8EncodedKeySpec(keyBytes));
+      return KeyFactory.getInstance(clientKeyAlgo).generatePrivate(new PKCS8EncodedKeySpec(keyBytes));
     } catch (InvalidKeySpecException e) {
       // Otherwise try PKCS8
       RSAPrivateCrtKeySpec keySpec = PKCS1Util.decodePKCS1(keyBytes);
-      return keyFactory.generatePrivate(keySpec);
+      return KeyFactory.getInstance(clientKeyAlgo).generatePrivate(keySpec);
```
Comment (manusa) [inline kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/internal/CertUtils.java:181]: The fix is brittle without a test that ensures this won't be reverted in the future.
A test would be good, but at least a comment pointing to the issue should be added here.

**Review by manusa — APPROVED** [review]

**Pull request merged**